### PR TITLE
Feature/add complex nodes

### PR DESCRIPTION
**Pull request by a-zakir**

### CODE REVIEW

File: src/solver/expressions/nodes/BinaryNode.cpp [hunk 1]
```diff
@@ -18,13 +18,14 @@
...
+#pragma once
+
+#include <antares/solver/expressions/nodes/BinaryNode.h>
 
 namespace Antares::Solver::Expressions
 {
-NegationNode::NegationNode(Node* n):
-    n_(n)
+BinaryNode::BinaryNode(Node* n1, Node* n2):
+    ParentNode<2>({n1, n2})
```
Comment (flomnes) [inline src/solver/expressions/nodes/BinaryNode.cpp:28]: 👍🏻 More convenient than the initializer list constructor

File: src/solver/expressions/include/antares/solver/expressions/nodes/ParentNode.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        if (children_.empty() || idx >= children_.size())
+        {
+            throw ParentNodeException(
+              "Antares::Solver::Expressions::ParentNode can't get the child node at position"
+              + std::to_string(idx));
+        }
+        else
+        {
+            return children_[idx];
+        }
```
Comment (flomnes) [inline src/solver/expressions/include/antares/solver/expressions/nodes/ParentNode.h]: Use `std::array<>::at` and catch exception `std::out_of_range` here ?

File: src/solver/hydro/management/HydroInputsChecker.cpp [hunk 3]
```diff
@@ -40,6 +40,7 @@ HydroInputsChecker::HydroInputsChecker(Antares::Data::Study& study):
     scenarioInitialHydroLevels_(study.scenarioInitialHydroLevels),
     scenarioFinalHydroLevels_(study.scenarioFinalHydroLevels)
 {
+    auto a = ERROR;
```
Comment (flomnes) [inline src/solver/hydro/management/HydroInputsChecker.cpp]: What gives ?

File: src/solver/expressions/include/antares/solver/expressions/nodes/ParentNode.h (hunk 2 above)
Comment (a-zakir) [reply src/solver/expressions/include/antares/solver/expressions/nodes/ParentNode.h]: in this case 2 exceptions will be thrown

File: src/tests/src/solver/expressions/test_expressions.cpp [hunk 4]
```diff
@@ -176,3 +176,88 @@ BOOST_FIXTURE_TEST_CASE(cloneVisitor_With_Add_Neg_ComponentVariableNode, Registr
...
+                          ParentNodeException,
+                          [&pos](const ParentNodeException& ex)
+                          {
+                              return strcmp(ex.what(),
+                                            (std::string("Antares::Solver::Expressions::ParentNode "
+                                                         "can't get the child node at position ")
+                                             + std::to_string(pos))
+                                              .c_str())
+                                     == 0;
+                          });
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_expressions.cpp:257]: IMO `BOOST_CHECK_THROW` would be enough here. You don't really need to check the error message

File: src/tests/src/solver/expressions/test_expressions.cpp [hunk 5]
```diff
@@ -176,3 +176,88 @@ BOOST_FIXTURE_TEST_CASE(cloneVisitor_With_Add_Neg_ComponentVariableNode, Registr
...
+BOOST_FIXTURE_TEST_CASE(subtraction_node, Registry<Node>)
+{
+    double num1 = 22.0, num2 = 8;
+    // num1+num2
+    Node* sub = create<SubtractionNode>(create<LiteralNode>(num1), create<LiteralNode>(num2));
+
+    PrintVisitor printVisitor;
+    const auto printed = printVisitor.dispatch(*sub);
+
+    BOOST_CHECK_EQUAL(printed, "(22.000000-8.000000)");
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_expressions.cpp:231]: More like
```suggestion
    BOOST_CHECK_EQUAL(printed, "22.000000-(8.000000)");
```

File: src/tests/src/solver/expressions/test_expressions.cpp [hunk 6]
```diff
@@ -176,3 +176,88 @@ BOOST_FIXTURE_TEST_CASE(cloneVisitor_With_Add_Neg_ComponentVariableNode, Registr
...
+    PrintVisitor printVisitor;
+    const auto printed = printVisitor.dispatch(*div);
+
+    BOOST_CHECK_EQUAL(printed, "22.000000/0.000000");
+    EvalVisitor evalVisitor;
+
+    BOOST_CHECK_EXCEPTION(evalVisitor.dispatch(*div),
+                          EvalVisitorDivisionException,
+                          [](const EvalVisitorDivisionException& ex)
+                          { return strcmp(ex.what(), "DivisionNode Division by zero") == 0; });
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_expressions.cpp:220]: Why not use `==` with two `std::strings` ?

File: src/tests/src/solver/expressions/test_expressions.cpp [hunk 7]
```diff
@@ -176,3 +176,88 @@ BOOST_FIXTURE_TEST_CASE(cloneVisitor_With_Add_Neg_ComponentVariableNode, Registr
...
+
+    BOOST_CHECK_EQUAL(printed, "22.000000/8.000000");
+    EvalVisitor evalVisitor;
+    BOOST_CHECK_EQUAL(evalVisitor.dispatch(*div), num1 / num2);
+}
+
+BOOST_FIXTURE_TEST_CASE(division_by_zero, Registry<Node>)
+{
+    double num1 = 22.0, num2 = 0.;
+    // num1+num2
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_expressions.cpp]:
```suggestion
    // num1/num2
```

File: src/tests/src/solver/expressions/test_expressions.cpp [hunk 8]
```diff
@@ -176,3 +176,88 @@ BOOST_FIXTURE_TEST_CASE(cloneVisitor_With_Add_Neg_ComponentVariableNode, Registr
...
+
+    BOOST_CHECK_EQUAL(printed, "22.000000*8.000000");
+    EvalVisitor evalVisitor;
+    BOOST_CHECK_EQUAL(evalVisitor.dispatch(*mult), num1 * num2);
+}
+
+BOOST_FIXTURE_TEST_CASE(division_node, Registry<Node>)
+{
+    double num1 = 22.0, num2 = 8;
+    // num1+num2
```
Comment (flomnes) [inline src/tests/src/solver/expressions/test_expressions.cpp]:
```suggestion
    // num1/num2
```

File: src/tests/src/solver/expressions/test_expressions.cpp (hunk 6 above)
Comment (a-zakir) [reply src/tests/src/solver/expressions/test_expressions.cpp:220]: same result

File: src/tests/src/solver/expressions/test_expressions.cpp (hunk 5 above)
Comment (a-zakir) [reply src/tests/src/solver/expressions/test_expressions.cpp:231]: if i accept your suggestion, we must add some () on some nodes print visitor:
s =sub(8,3)= 8-(3)
d =div(s,3) =  8-(3)/3

**Pull request merged**